app/src/main/java/org/mozilla/fenix/components/Core.kt

```diff
@@ -146,13 +148,18 @@ class Core(private val context: Context, private val crashReporter: CrashReporti
                 MediaMiddleware(context, MediaService::class.java),
                 DownloadMiddleware(context, DownloadService::class.java),
                 ReaderViewMiddleware(),
-                ThumbnailsMiddleware(thumbnailStorage)
+                ThumbnailsMiddleware(thumbnailStorage),
+                UndoMiddleware(::lookupSessionManager, context.getUndoDelay())
             ) + EngineMiddleware.create(engine, ::findSessionById)
         ).also {
             it.dispatch(RecentlyClosedAction.InitializeRecentlyClosedState)
         }
     }
 
+    private fun lookupSessionManager(): SessionManager {
+        return sessionManager
```

Comment:
Is this to get around init loops? 😬

Reply:
Yes. Super ugly. I hope SessionManager can die soon.